src/Storage/Builder.php

```diff
@@ -0,0 +1,189 @@
+{
+    /**
+     * The token header
+     *
+     * @var array
+     */
+    private $headers = ['typ'=> 'JWT', 'alg' => 'none'];
+
+    /**
+     * The token claim set
+     *
+     * @var array
+     */
+    private $claims = [];
+
+    /**
+     * The data encoder
+     *
+     * @var Parsing\Encoder
+     */
+    private $encoder;
+
+    /**
+     * Initializes a new builder
+     */
+    public function __construct(Parsing\Encoder $encoder)
+    {
+        $this->encoder = $encoder;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function canOnlyBeUsedBy(string $audience, bool $addHeader = false): BuilderInterface
+    {
+        $audiences = $this->claims['aud'] ?? [];
+
+        if (!in_array($audience, $audiences)) {
+            $audiences[] = $audience;
+        }
+
+        return $this->setRegisteredClaim('aud', $audiences, $addHeader);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function expiresAt(int $expiration, bool $addHeader = false): BuilderInterface
+    {
+        return $this->setRegisteredClaim('exp', $expiration, $addHeader);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function identifiedBy(string $id, bool $addHeader = false): BuilderInterface
+    {
+        return $this->setRegisteredClaim('jti', $id, $addHeader);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function issuedAt(int $issuedAt, bool $addHeader = false): BuilderInterface
+    {
+        return $this->setRegisteredClaim('iat', (int) $issuedAt, $addHeader);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function issuedBy(string $issuer, bool $addHeader = false): BuilderInterface
+    {
+        return $this->setRegisteredClaim('iss', $issuer, $addHeader);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function canOnlyBeUsedAfter(int $notBefore, bool $addHeader = false): BuilderInterface
+    {
+        return $this->setRegisteredClaim('nbf', $notBefore, $addHeader);
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function relatedTo(string $subject, bool $addHeader = false): BuilderInterface
+    {
+        return $this->setRegisteredClaim('sub', $subject, $addHeader);
+    }
+
+    /**
+     * Configures a registered claim
+     */
+    private function setRegisteredClaim(string $name, $value, bool $addHeader): BuilderInterface
+    {
+        $this->with($name, $value);
+
+        if ($addHeader) {
+            $this->headers[$name] = $this->claims[$name];
+        }
+
+        return $this;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function withHeader(string $name, $value): BuilderInterface
+    {
+        $this->headers[$name] = $value;
+
+        return $this;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function with(string $name, $value): BuilderInterface
+    {
+        $this->claims[$name] = $value;
+
+        return $this;
+    }
+
+    private function encode(array $items): string
```

Comment:
Not sure... sometimes you don't want to encode with JSON+Base64Url (binary data is just base64url for example)